azure-iot-device/tests/common/pipeline/test_pipeline_stages_mqtt.py

```diff
@@ -159,168 +222,219 @@ def test_sets_transport_attribute_on_root(self, stage, transport, op_set_connect
         stage.run_op(op_set_connection_args)
         assert stage.previous.transport == transport.return_value
 
-    @pytest.mark.it("Completes with success if no exception")
+    @pytest.mark.it("Sets the pending connection op tracker to None")
+    def test_pending_conn_op(self, stage, transport, op_set_connection_args):
+        stage.run_op(op_set_connection_args)
+        assert stage._pending_connection_op is None
+
+    @pytest.mark.it("Completes the operation with success, upon successful execution")
     def test_succeeds(self, stage, transport, op_set_connection_args):
         stage.run_op(op_set_connection_args)
         assert_callback_succeeded(op=op_set_connection_args)
 
-    @pytest.mark.it("Completes with failure on exception")
-    def test_fails_on_exception(self, stage, transport, op_set_connection_args, mocker):
-        transport.return_value = None
-        stage.run_op(op_set_connection_args)
-        assert_callback_failed(op=op_set_connection_args)
 
+@pytest.mark.describe("MQTTTransportStage - .run_op() -- called with ConnectOperation")
+class TestMQTTProviderExecuteOpWithConnect(RunOpTests):
+    @pytest.mark.it("Sets the ConnectOperation as the pending connection operation")
+    def test_sets_pending_operation(self, stage, create_transport, op_connect):
+        stage.run_op(op_connect)
+        assert stage._pending_connection_op is op_connect
+
+    @pytest.mark.it("Cancels any already pending connection operation")
+    @pytest.mark.parametrize(
+        "pending_connection_op",
+        [
+            pytest.param(pipeline_ops_base.ConnectOperation(), id="Pending ConnectOperation"),
+            pytest.param(pipeline_ops_base.ReconnectOperation(), id="Pending ReconnectOperation"),
+            pytest.param(pipeline_ops_base.DisconnectOperation(), id="Pending DisconnectOperation"),
+        ],
+    )
+    def test_pending_operation_cancelled(
```

Comment:
>test_pending_operation_cancelled [](start = 8, length = 32)

this is the behavior that matters.  tests_sets_pending_operation and test_pending_conn_op are testing implementation details that support this functionality.  I think this test alone is sufficient and those other 2 are not necessary.

Reply:
oh, wait, you have one test that verifies that _pending_connection_op gets set, and another test that verifies that the pending op gets cancelled.  I can see why you broke these up, but I still don't like how the private attribute leaks into the test.  I don't like it, but I don't think it's worth obsessing over either.  

---
In reply to: [311657202](https://github.com/Azure/azure-iot-sdk-python-preview/pull/159#discussion_r311657202) [](ancestors = 311657202)

Reply:
I could see the argument to not having a specific test (i.e. requirement) for things related to private attributes (i.e. fold these tests together), but the private attribute does have to be at asserted upon somewhere, otherwise we don't know it's being set/unset correctly.

---
In reply to: [311658611](https://github.com/Azure/azure-iot-sdk-python-preview/pull/159#discussion_r311658611) [](ancestors = 311658611,311657202)